src/util/base.cc

```diff
@@ -0,0 +1,231 @@
+#include <assert.h>
+#include "core/util/base64.h"
+
+namespace bpo = boost::program_options;
+
+#define BASE32_BUF_SIZE 400
+#define BASE64_BUF_SIZE 120
+
+static_assert(BASE32_BUF_SIZE % 40 == 0, "Invalid BASE32_BUF_SIZE");
+static_assert(BASE64_BUF_SIZE % 12 == 0, "Invalid BASE64_BUF_SIZE");
+
+BaseCommand::BaseCommand() : m_desc("General options")
+{
+  m_desc.add_options()("help,h", "produce this help message")(
+      "type,t", bpo::value<std::string>(&m_opt_type), "encode/decode")(
+      "infile,i", bpo::value<std::string>(&m_opt_infile), "input file")(
+      "outfile,o", bpo::value<std::string>(&m_opt_outfile), "output file");
+}
+
+void BaseCommand::PrintUsage(const std::string& name) const
+{
+  std::cout << "Syntax: " << name << " <options> encode <inFile> <outfile>"
+            << std::endl;
+  std::cout << "or    : " << name << " <options> decode <inFile> <outfile>"
+            << std::endl;
+  std::cout << std::endl;
+  std::cout << m_desc << std::endl;
+}
+
+template <typename ToBase, typename ToByte, typename SizeConst>
+bool process(
+    ToBase ByteStreamToBase,
+    ToByte BaseToByteStream,
+    SizeConst Size,
+    bool encode,
+    std::istream& input,
+    std::ostream& output)
+{
+  std::uint8_t in_buffer[Size + 1];
+  std::uint8_t out_buffer[Size * 2 + 1];
+  size_t ret(0);
+  memset(in_buffer, 0, sizeof(in_buffer));
+  memset(out_buffer, 0, sizeof(out_buffer));
+  do
+    {
+      input.read(reinterpret_cast<char*>(in_buffer), Size);
+      if (input.bad())
+        {
+          std::cerr << "Error while reading input ! " << strerror(errno)
+                    << std::endl;
+          return false;
+        }
+
+      if (encode)
+        ret = ByteStreamToBase(
+            in_buffer,
+            input.gcount(),
+            reinterpret_cast<char*>(out_buffer),
+            Size * 2);
+      else
+        ret = BaseToByteStream(
+            reinterpret_cast<char*>(in_buffer),
+            input.gcount(),
+            out_buffer,
+            Size * 2);
+
+      output.write(reinterpret_cast<char*>(out_buffer), ret);
+      if (output.bad())
+        {
+          std::cerr << "Error : Output to stream failed !" << std::endl;
+          return false;
+        }
+      output.flush();
+    } while (input.good());
+  return true;
+}
+
+bool BaseCommand::Impl(
+    const std::string& cmd_name,
+    int argc,
+    const char* argv[])
+{
+  bpo::variables_map vm;
+  bpo::positional_options_description pos;
+  pos.add("type", 1).add("infile", 1).add("outfile", 1);
+  try
+    {
+      bpo::parsed_options parsed = bpo::command_line_parser(argc, argv)
+                                       .options(m_desc)
+                                       .positional(pos)
+                                       .run();
+      bpo::store(parsed, vm);
+      bpo::notify(vm);
+    }
+  catch (const bpo::error& e)
+    {
+      std::cerr << GetName() << " ERROR: " << e.what() << std::endl;
+      return false;
+    }
```

Comment:
For all `try` blocks, please consider using (and getting in the habit of using elsewhere) our exception dispatcher in `src/core/util/exception.h`. Implementation examples are scattered throughout the code or I can provide you one if needed. You can also add to the dispatcher (as you may want to do here) if you wish.